SIMSBIOHUB-188 Manage Animals

MacDeluca opened the pull request:
## Links to Jira Tickets

-[SIMSBIOHUB-188](https://apps.nrs.gov.bc.ca/int/jira/browse/SIMSBIOHUB-188)

## Description of Changes

- Overhauled the Edit Animal modal to be a full screen page instead. Accessed through the Manage Animals button on a survey.
- Animals in this view appear listed in the sidebar by their alias. Clicking on one of these items opens an accordion to the various form sections.
- Each form section has been adapted into a list of static cards. You can edit or delete cards from the relevant button icons. You can also add another card from the top of the page. Clicking the add or edit buttons brings up a form modal that will send a request on submission of the form.
- Deleting any card should prompt a confirm delete modal.
- Telemetry devices still work slightly differently than the rest of this page in that device and deployments get added together and deployments can be deleted but devices cannot. It's fairly consistently integrated otherwise.
- Currently all sections hit the bulk endpoint in Critterbase (allows multiple sections to be added at once). This can be modified to single endpoints per each form section if this is the final version of this UI.

## Testing Notes

- A small patch to critterbase was required to fix a bug in the capture delete function. This should be in live dev already.
- Currently some tests are broken but waiting till design / flow is more finalized before fixing

jeznorth committed:
WIP

curtisupshall committed:
SIMSBIOHUB-226: Map improvements

curtisupshall committed:
Merge branch 'survey-prototype' of github.com:bcgov/biohubbc into survey-prototype

jeznorth committed:
UI Updates

curtisupshall committed:
SIMSBIOHUB-226: Use square bounding box

curtisupshall committed:
Merge branch 'survey-prototype' of github.com:bcgov/biohubbc into survey-prototype

curtisupshall committed:
Merge branch 'dev' into survey-prototype

curtisupshall committed:
SIMSBIOHUB-226: Added table loading skeleton

curtisupshall committed:
Merge branch 'dev' into survey-prototype

jeznorth committed:
UI Tweaks and Formatting

jeznorth committed:
UI Updates

MacDeluca committed:
merged survey-prototype

MacDeluca committed:
basic page layout created for animals

MacDeluca committed:
animal side bar boilerplate

MacDeluca committed:
rendering the animal data inside the sidebar

MacDeluca committed:
created sub list for animals

MacDeluca committed:
made list items buttons

MacDeluca committed:
removed SurveyObservationHeader and replaced with generic SurveySectionHeader

MacDeluca committed:
refactored similar page layouts into single component

GrahamS-Quartech committed:
Simple prototype for having the sidebar scroll to various headers within the main animal form view on the new full screen animal page

GrahamS-Quartech committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of github.com:bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

GrahamS-Quartech committed:
Plugged in AddEditAnimal to main component for fullscreen animal view

GrahamS-Quartech committed:
untracked file woops

MacDeluca committed:
added state for selected critters / sections

GrahamS-Quartech committed:
Configured AddEditAnimal to receieve props for critter id and selected form element

MacDeluca committed:
updated constants

MacDeluca committed:
fixed broken AnimalList

GrahamS-Quartech committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of github.com:bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

GrahamS-Quartech committed:
Added save button, removed divider, added editing critter header

GrahamS-Quartech committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of github.com:bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

MacDeluca committed:
updated button from UI to connect to animal page

GrahamS-Quartech committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of github.com:bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

MacDeluca committed:
added fix for auto closing accordion sections

MacDeluca committed:
updated state for selected critter

GrahamS-Quartech committed:
Form can save now, loading button implemented correctly

MacDeluca committed:
commit before critter url param

GrahamS-Quartech committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of github.com:bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

GrahamS-Quartech committed:
Fix for measurementanimalform not always using the correct field type on initial render

GrahamS-Quartech committed:
snack bar added for success and error case, basic cross section error message, added consistent critter sorting to sidebar

MacDeluca committed:
passing the new params correctly

MacDeluca committed:
finalized the URL params

MacDeluca committed:
fixed merge conflicts

MacDeluca committed:
fixed some issues with url param needing to be parsed to int

MacDeluca committed:
accordion sections now stay highlighted

MacDeluca committed:
accordion sections now stay highlighted

GrahamS-Quartech committed:
Refactored Formik component structure so that both AddEditAnimal and AnimalList can access shared formik state variables

MacDeluca committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

MacDeluca committed:
updated ui layout and small bug fix for survey page

GrahamS-Quartech committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of github.com:bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

GrahamS-Quartech committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of github.com:bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

GrahamS-Quartech committed:
Minor refactor to remove duplicated code across some of the fullscreen animal components

MacDeluca committed:
discard changes now works correctly

GrahamS-Quartech committed:
Implemented add animal button and dialog

GrahamS-Quartech committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of github.com:bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

MacDeluca committed:
yes no dialog working correctly for switching between critters

MacDeluca committed:
yes no dialog working correctly for switching between critters

MacDeluca committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

curtisupshall committed:
SIMSBIOHUB-226: Survey Prototyping + Observation Map View (#1116)

* Various survey page layout changes
* Adds the ability to see added survey observations in a map view

curtisupshall committed:
Merge branch 'dev' into dev-staging

GrahamS-Quartech committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of github.com:bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

GrahamS-Quartech committed:
Fixed an issue with the measurement schema where 0 would not be permitted as a quantitative value

GrahamS-Quartech committed:
Fixed regression causing save button to stop responding

MacDeluca committed:
Added critter-id section in form

MacDeluca committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

MacDeluca committed:
added critter ID with copy feature

MacDeluca committed:
added critter ID with copy feature

MacDeluca committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

MacDeluca committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

MacDeluca committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

MacDeluca committed:
pre commit before add btn refactor

GrahamS-Quartech committed:
Basic implementation of telemetry devices within full screen animal form. Loads and displays correctly but needs more work to save.

GrahamS-Quartech committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of github.com:bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

MacDeluca committed:
refactored capture

MacDeluca committed:
refactored the general form section

jeznorth committed:
UI Fixes

GrahamS-Quartech committed:
Implemented saving and editing deployments and devices to the backend in the new card view layout

MacDeluca committed:
finalized add btn refactor + bug fix for stable IDs'

MacDeluca committed:
passed [] to device in all test suites

curtisupshall committed:
Merge branch 'dev' into dev-staging

NickPhura committed:
Fix issues

curtisupshall committed:
Merge branch 'dev' into dev-staging

curtisupshall committed:
Merge branch 'dev-staging' of github.com:bcgov/biohubbc into dev-staging

MacDeluca committed:
Started boilerplate for small card display

GrahamS-Quartech committed:
Progress in making a read only card view for the fullscreen animals. Currently only capture has the base level of UI functionality.

GrahamS-Quartech committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of github.com:bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

MacDeluca committed:
commit before merge

MacDeluca committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

NickPhura committed:
Fix pipeline route issues

NickPhura committed:
Fix non-unique key issue

MacDeluca committed:
basic section cards created

GrahamS-Quartech committed:
Modifying existing subsections so that we can extract individual content components from each of them

GrahamS-Quartech committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of github.com:bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

MacDeluca committed:
passing critter to AnimalSectionDataCards

MacDeluca committed:
now handling undefined values inside the cards

MacDeluca committed:
renamed component / moved to EditDeleteStubCard.tsx

GrahamS-Quartech committed:
Checkpoint before I start ripping all the telemetry api stuff into different components

GrahamS-Quartech committed:
device deployment addition edit and removal appears to be correctly implemented now, needs a bit more testing though

GrahamS-Quartech committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of github.com:bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

MacDeluca committed:
formatted stub data card subheader and added icons for sections

MacDeluca committed:
fixed merge conflict around edit dialogs

NickPhura committed:
Merge branch 'dev' into dev-staging

GrahamS-Quartech committed:
Re-added family section without the infinite render loop

MacDeluca committed:
fixed issue with general section not rendering dialog

GrahamS-Quartech committed:
refreshFamilies update with critterData instead

GrahamS-Quartech committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of github.com:bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

MacDeluca committed:
fixed grid container issue with form components

GrahamS-Quartech committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of github.com:bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

MacDeluca committed:
updated styling for critter id

MacDeluca committed:
improvement for state flow of adding new records

jeznorth committed:
SIMSBIOHUB-258 (#1134)

* Minor Fixes / UI Tweaks
* Fix permit component list key
* Rename old sampling method form to sampling strategy

Co-authored-by: Nick Phura <nickphura@gmail.com>

MacDeluca committed:
updated styling in AddEditAnimal

GrahamS-Quartech committed:
Made family labels available to the card view by just calling the hook, can't do it the same way as the other parts without chaning the cb api. Fixed some issues with creating new families under this format as well.

GrahamS-Quartech committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of github.com:bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

MacDeluca committed:
updated tests to handle new data formats

GrahamS-Quartech committed:
Restored async form validation to telemetry form in SurveyAnimalPage

GrahamS-Quartech committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of github.com:bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

MacDeluca committed:
created delete modal for sections

MacDeluca committed:
updated styling for Grid container

GrahamS-Quartech committed:
Changes are now saved when modal closes instead of using save and discard buttons. Fixed async telemetry validation so that it does not fire in the background when using other forms. Also fixed a bug where deleting the last deployment would crash the app.

GrahamS-Quartech committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of github.com:bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

MacDeluca committed:
fixed annoying issue with transitions not firing correctly

MacDeluca committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

MacDeluca committed:
fixed merge conflict issue with imports

GrahamS-Quartech committed:
Fixed some bugs with saving measuremnts and saving location changes for mortality

GrahamS-Quartech committed:
Lifted redundant allFamilies request out of data cards, instead get from parent. Now it will refresh when form is submitted, but not when switching sections or other interactions.

MacDeluca committed:
updated AddEditAnimal

curtisupshall committed:
SIMSBIOHUB-296/313: Import Observations (#1128)

* Add's UIs for uploading CSV files to survey observations
* Adds a new database table to track survey observation CSV submissions, survey_observation_submission
* Alters the survey_observation table to make optional the sampling site, method, and period columns
* Adds service and repo methods for processing observation CSVs

GrahamS-Quartech committed:
Replaced the placeholder text for device section

GrahamS-Quartech committed:
Removed debug console logs and extracted setPopup to be reusable

al-rosenthal committed:
SIMSBIOHUB-177: Survey Study Area enhancements  (#1136)

* Survey Study area can now import files with multiple features
* Drawn features can be created and saved to a survey
* New card for displaying and editing map features
* Region selector updated

NickPhura committed:
Merge branch 'dev' into dev-staging

MacDeluca committed:
updated dialog styling

MacDeluca committed:
fixed TelemetryDeviceForm merge conflict

MacDeluca committed:
updated telemetry modal

MacDeluca committed:
updated telemetry modal

MacDeluca committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

MacDeluca committed:
fixed conflicts with dev-staging

MacDeluca committed:
updated styling in SurveyFullPageLayout

MacDeluca committed:
fixed formatting and linter issues

MacDeluca committed:
fixed formatting issues and sonar clouds

MacDeluca committed:
refactored verbose function

GrahamS-Quartech committed:
Fixed incorrect errors occuring sometimes when adding or deleting deployments due to the cached device details used for the async validation not being refreshed. Also prevented the add deployment action from including empty values in the result so that it will not overwrite any existing metadata that may be present for an existing device id.

GrahamS-Quartech committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of github.com:bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

curtisupshall commented on app/src/utils/UtilsJSX.tsx:
I would move the contents of this file to `Utils.ts` and just rename the file to `Utils.tsx`.

MacDeluca replied:
Fixedarooni

curtisupshall commented on app/src/utils/UtilsJSX.tsx:
```diff
@@ -0,0 +1,19 @@
+ * @param message string to show
+ * @param context reference to current DialogContext
+ */
+export const setPopup = (message: string, context: IDialogContext) => {
```
I would probably rename this to `setSnackbar(message, dialogContext)`, or something to that effect; `setPopup` makes me think of a dialog rather than a snackbar

MacDeluca replied:
Done!

curtisupshall commented on app/src/features/surveys/view/survey-animals/form-sections/MeasurementAnimalForm.tsx:
```diff
@@ -98,23 +86,31 @@ interface MeasurementFormContentProps {
     handleChange(event);
     setFieldValue(valueName, '');
     setFieldValue(optionName, '');
     const m = measurements?.find((m) => m.taxon_measurement_id === event.target.value);
```
We should avoid single-letter variable names; I would call this `const measurement`

curtisupshall replied:
Since `measurement` is already taken, you could rename that to `const [currentMeasurement, setCurrentMeasurement]` or maybe `const [formMeasurement, setFormMeasurement]`

MacDeluca replied:
fixed

curtisupshall commented on app/src/features/surveys/view/survey-animals/form-sections/MeasurementAnimalForm.tsx:
```diff
@@ -98,23 +86,31 @@ interface MeasurementFormContentProps {
-  const tMeasurementIDName = getAnimalFieldName<IAnimalMeasurement>(NAME, 'taxon_measurement_id', index);
-  const valueName = getAnimalFieldName<IAnimalMeasurement>(NAME, 'value', index);
-  const optionName = getAnimalFieldName<IAnimalMeasurement>(NAME, 'qualitative_option_id', index);
+  const tMeasurementIDName = getAnimalFieldName<IAnimalMeasurement>(animalKeyName, 'taxon_measurement_id', index);
```
avoid abbreviations that obfuscate the purpose of a variable (`tMeasurement` could be "time measurement", "telemetry measurement", ...)

MacDeluca replied:
fixed

curtisupshall commented on app/src/features/surveys/components/EditDeleteStubCard.tsx:
```diff
@@ -0,0 +1,68 @@
+      if (value == null || value === '') {
+        return;
+      }
+      formatArr.push(`${key} • ${value}`);
```
Nit: prefer [HTML entities](https://www.w3schools.com/html/html_entities.asp) over Unicode.

curtisupshall replied:
(If you can't get this to work very easily, just ignore this comment)

MacDeluca replied:
Yea the solution for this is a bit more complex than I think its worth. Just gonna keep it like this for now

curtisupshall commented on app/src/features/surveys/components/SurveySectionHeader.tsx:
```diff
@@ -0,0 +1,53 @@
+  title: string;
+}
+
+const SurveySectionHeader: React.FC<SurveySectionHeaderProps> = (props) => {
```
Remove `React.FC`

MacDeluca replied:
Fixed 💯

MacDeluca replied:
Fixed

curtisupshall commented on app/src/features/surveys/view/SurveyAnimals.tsx:
```diff
@@ -510,7 +511,7 @@ const SurveyAnimals: React.FC = () => {
         dialogLoading={isSubmittingTelemetry}
         component={{
-          element: <TelemetryDeviceForm removeAction={handleRemoveDeployment} mode={telemetryFormMode} />,
+          element: <TelemetryDeviceForm removeAction={handleRemoveDeployment} survey_critter_id={-1} />, // This section will be removed
```
If there is an existing ticket to track the work of removing this section, please link it :)

MacDeluca replied:
@GrahamS-Quartech?

GrahamS-Quartech replied:
I'm not sure what the ghost of Graham past meant by this but it's probably appropriate to remove all / most of SurveyAnimals.tsx considering much of it is now copied into the various new components that make up this ticket.

curtisupshall commented on app/src/features/surveys/view/survey-animals/animal.ts:
```diff
@@ -6,6 +6,8 @@ import { v4 } from 'uuid';
 import { AnyObjectSchema, InferType, reach } from 'yup';
 import { AnimalTelemetryDeviceSchema } from './device';
 
+export const MANAGE_ANIMALS_DEFAULT_URL_PARAM = '0';
```
Is there a lot of value in tracking this as a constant?

MacDeluca replied:
Removed not needed anymore

curtisupshall commented on app/src/features/surveys/view/SurveyAnimals.tsx:
```diff
@@ -535,14 +536,25 @@ const SurveyAnimals: React.FC = () => {
+        </Typography>
+        <Button
+          component={RouterLink}
+          to={`animals/${MANAGE_ANIMALS_DEFAULT_URL_PARAM}`}
```
I think I understand what this code is for... On the `/animals` page, if it receives a selected critter ID of exactly `0`, then the page renders in some default state; and if instead the page receives some actual critter ID, then the user is able to manage details for that particular critter?

I think passing an out-of-bounds critter ID for this purpose is an anti-pattern. Instead of using a path param, e.g. `/animals/{id}`, I would use a query param instead, which would allow it to be optional. E.g., `/animals` and `/animals?cid=12`

MacDeluca replied:
I prefer this stratgey, implementing this change now.

MacDeluca replied:
Fixed

curtisupshall commented on app/src/features/surveys/view/survey-animals/AnimalList.tsx:
```diff
@@ -0,0 +1,158 @@
+  onAddButton: () => void;
+}
+
+const AnimalList = ({ selectedSection, onSelectSection, critterData, isLoading, onAddButton }: AnimalListProps) => {
```
Nit: Avoid destructuring large props lists

MacDeluca replied:
Fixed :)

MacDeluca replied:
Fixed

MacDeluca committed:
Fixed issue with Captures / Mortality map not rendering points

MacDeluca committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

jeznorth committed:
UI Clean-up - Animal List

MacDeluca committed:
minor tweaks / suggestions

MacDeluca committed:
switched from path param to query param for critter ID.

MacDeluca committed:
handled PR measurement suggestions

MacDeluca committed:
Moved contents from UtilsJSX.tsx -> Utils.tsx

MacDeluca committed:
changed setPopup -> setMessageSnackbar

MacDeluca committed:
reverted observations related useEffect hook dependencies, now same as dev-staging

jeznorth committed:
Adding Taxon and Sex to Animal List Item

MacDeluca committed:
updated page -> edit critter flow

MacDeluca committed:
updated subHeaderData type to handle Ecological Units more appropriately

curtisupshall commented on app/src/features/surveys/components/EditDeleteStubCard.tsx:
```diff
@@ -0,0 +1,68 @@
+import { grey } from '@mui/material/colors';
+import React from 'react';
+
+export type ISubHeaderData = Record<string, string | number | undefined> | string;
```
Nit: this type is not an interface, so you can just call this `SubHeaderData`.

I might reconsider accepting an `object | string`-like React prop, since the resulting UI behvaiour varies quite a bit. Two things to consider:

1. Use `type SubHeaderData = { key: string; value?: string | number }`. This way, if I want my header to just display a single string, I can pass `header={{ key: 'My Header }}`.
2. Inline your header rendering method:
```javascript
<CardHeader
  subheader={
    props.header.map((header) => [header.key, header.value].filter(Boolean).join(' • ')).join(' | ')
  }
  ...
/>

curtisupshall replied:
If you find this strategy compelling, you may reconsider the property names for `SubHeaderData`, e.g. `title` instead of `key`

MacDeluca replied:
Fixed the type.

It might almost be beneficial to only accept strings for the subHeaderData. 
Allow the subheader format to be handled by the parent component for more flexibility.

Just from exploring this solution, the objects passed to subHeaderData become very verbose.

MacDeluca committed:
deprecated menu controls for individual animals

JeremyQuartech committed:
SIMSBIOHUB-336/337 (#1142)

* Adds Sampling Site record count to title area, Example: Sampling Sites (10)
* Adds Observation record count to title area of observations map component and manage observations page, Example: Observations (10)
* Adds 'supplementaryObservationData' section to get survey observations endpoint. This allows observation record counts to be displayed, even if the get observations response becomes paginated or virtualized at a later date

GrahamS-Quartech committed:
SIMSBIOHUB 343 - Incorrect date overlap error in sampling methods (#1144)

* Fixed an issue where the yup schema used for editing sampling methods used the wrong field for a date overlap check.
* Removed an unintended required().
* Added a refresh call when submitting sampling method changes so they are reflected in the UI. Otherwise you will still see the previous value if you go back to the sampling methods view even though the new value did save to the db.

curtisupshall committed:
SIMSBIOHUB-334/350/352 (#1143)

* Added snackbar popup upon successfully saving observations.
* Removed single-click to edit; Fixed bug that made row go into edit mode when clicking row delete icon.
* Added seconds to the observation time

MacDeluca committed:
only rendering critter_id in general section

MacDeluca committed:
updated to full width styling for animal forms

MacDeluca committed:
updated to now say "Save" on updating form section

MacDeluca committed:
fixed transition issue with animal cards

jeznorth committed:
UI WIP

jeznorth committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

MacDeluca committed:
updated EditDeleteStubCard to now accept strings vs objects for the subHeader prop

MacDeluca committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

MacDeluca committed:
added empty state placeholder

MacDeluca committed:
empty state bug fix

jeznorth committed:
UI WIP

MacDeluca committed:
Updated empty state logic for manage animals

MacDeluca committed:
updated empty state change and handled merge conflict

MacDeluca committed:
moved add button below cards + removed dialog info text

MacDeluca committed:
removed old deprecated animal code

jeznorth committed:
UI WIP

jeznorth committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

JeremyQuartech committed:
SIMSBIOHUB-323: Remove Ecological Season (#1153)

* Removes all UI references to Ecological Season
* Removes all API references to Ecological Seasons
* Adds DB migration for deprecation comments on Ecological seasons column and xref table
* Removes ecological season from DB seed

jeznorth committed:
Form Updates

MacDeluca committed:
fixed mortality bug

MacDeluca committed:
fixed merge conflict

GrahamS-Quartech committed:
SIMSBIOHUB 347/341 - Names from shapefiles (#1146)

* Added some additional logic to the API to extract a name and description from the provided shape file when creating sampling sites.
* It will search the keys of the geometry's properties for "name" or "label" and "des" or "desc" or "descr" (exact match), then use the value for the sampling site name and description respectively.
* Added a character limit to the sampling site name field in the edit form.

---------

Co-authored-by: JeremyQuartech <jeremy.kissack@quartech.com>

jeznorth committed:
UI Updates

al-rosenthal committed:
TechDebt: Map container refactor (#1148)

Map container removed in lieu of using the leaflet component 
Map css moved into single component
Removed project location data  

---------

Co-authored-by: Nick Phura <nickphura@gmail.com>
Co-authored-by: JeremyQuartech <jeremy.kissack@quartech.com>
Co-authored-by: jeznorth <twenty15creative@gmail.com>

al-rosenthal committed:
SIMSBIOHUB-353/325: Survey form updates (#1149)

Removed field method and site selection strategies are now required on surveys

MacDeluca committed:
bug fix for loading state of animal sections, snackbar message now firing after animation completes

KjartanE committed:
SIMSBIOHUB-342:  Delete Sampling Site (#1147)

* Add delete sample site button to menu drop down
* Check for any associated observations to sample site.
* hook up api sample-site delete function.

* SIMSBIOHUB-336/337 (#1142)

* Adds Sampling Site record count to title area, Example: Sampling Sites (10)
* Adds Observation record count to title area of observations map component and manage observations page, Example: Observations (10)
* Adds 'supplementaryObservationData' section to get survey observations endpoint. This allows observation record counts to be displayed, even if the get observations response becomes paginated or virtualized at a later date

* SIMSBIOHUB 343 - Incorrect date overlap error in sampling methods (#1144)

* Fixed an issue where the yup schema used for editing sampling methods used the wrong field for a date overlap check.
* Removed an unintended required().
* Added a refresh call when submitting sampling method changes so they are reflected in the UI. Otherwise you will still see the previous value if you go back to the sampling methods view even though the new value did save to the db.

* SIMSBIOHUB-334/350/352 (#1143)

* Added snackbar popup upon successfully saving observations.
* Removed single-click to edit; Fixed bug that made row go into edit mode when clicking row delete icon.
* Added seconds to the observation time

* Clean up functions and excess data calls

* fix tests

---------

Co-authored-by: JeremyQuartech <123425360+JeremyQuartech@users.noreply.github.com>
Co-authored-by: GrahamS-Quartech <112989452+GrahamS-Quartech@users.noreply.github.com>
Co-authored-by: Curtis Upshall <curtisupshall@gmail.com>

MacDeluca committed:
updated forms from firing onBlur to onChange

MacDeluca committed:
updated collectionUnit and family tests

jeznorth committed:
UI WIP

jeznorth committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

MacDeluca committed:
corrected code smells

MacDeluca committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

JeremyQuartech commented on app/src/features/surveys/view/survey-animals/TelemetryDeviceForm.tsx:
```diff
@@ -158,12 +175,10 @@ const DeviceFormSection = ({ values, index, mode, removeAction }: IDeviceFormSec
         <Grid container spacing={3}>
           <Grid item xs={6}>
-            <CustomNumberField
+            <CustomTextField
```
It seems that this field isn't always validating properly. Sometimes when you enter a new value, it doesn't if the device already exists and will allow a submission resulting in 500 error. Another thing that could be fixed is that when editing a device / deployment, this field is enabled, but maybe it should be disabled like with device id.

JeremyQuartech commented on app/src/features/surveys/view/survey-animals/TelemetryDeviceForm.tsx:
```diff
@@ -204,13 +219,13 @@ const DeviceFormSection = ({ values, index, mode, removeAction }: IDeviceFormSec
           <Grid item xs={6}>
             <TelemetrySelectField
```
The keyx / cfg upload functionality for Vectronic and Lotek devices has disappeared when adding a new device, but still shows up on edit

GrahamS-Quartech committed:
SIMSBIOHUB 324 - Multiple Ecological Variables (Intended Outcomes) (#1152)

* Added a migration that removes the intended_outcome_id column from the survey table and shifts the responsibility onto a join table called survey_intended_outcome. This makes it possible to have multiple intended outcomes associated with a single survey.
* Modified the frontend and API to support this change. The dropdown in the add and edit forms for the survey have been replaced with a multiselect. Multiselect will still display the subtext associated with these elements.
* In various cases where a single intended outcome was displayed before, they display as a comma deliminated list instead.
* These items now display with the label of "Ecological Variables" in the frontend survey components.

JeremyQuartech commented on app/src/features/surveys/view/survey-animals/form-sections/MortalityAnimalForm.tsx:
Save button is unresponsive

JeremyQuartech replied:
Seems this is an issue on all but the telemetry form... presumably this is a known issue already

GrahamS-Quartech committed:
SIMSBIOHUB 335: Tab Through Observation Edit (#1156)

* Modified each custom renderEditCell component of the ObservationTable to manually set hasFocus. This allows the user to Tab forward to the next cell or Shift+Tab back to the previous.
* Wrapped some of the simpler TextField / DatePicker renders in their own functional components so that the same hasFocus pattern could be used.

MacDeluca committed:
added optional sorting for cbSelect

jeznorth committed:
UI WIP

jeznorth committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

NickPhura committed:
SIMSBIOHUB-356: Update survey list endpoint to return a subset of survey fields (#1158)

* Replace existing survey list endpoint with index, which returns a subset of survey fields rather than all fields.

MacDeluca committed:
quick format of capture form

NickPhura committed:
SIMSBIOHUB-331: Keycloak Library Updates (#1151)

* Swap out old keycloak libraries for newer more up to date libraries.
* Add RouteGuards tests.
* Add git action to clean PR builds for PRs against non-base branches

GrahamS-Quartech commented:
I understand this is still a WIP so apologies if these are well known but I did a quick run through and noticed a few things:

- Formatting / spacing on family modal seems incorrect compared to the other sections.
- Dragging the map around will place the marker, I don't believe it did this previously.
- Doesn't seem like there's any way to place the release marker on the map anymore? Maybe this still needs design work.
- Triggering error messages on the Add Telemetry modal will cause the dialog box to resize horizontally.

jeznorth committed:
UI WIP

MacDeluca committed:
added controls for markers

MacDeluca committed:
fixed merge conflict

jeznorth committed:
Mortality Form Updates

jeznorth committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

MacDeluca committed:
dissabling controls after selecting marker on map

MacDeluca committed:
updated fieldset titles

MacDeluca committed:
updated pcod/ucod to render asc

MacDeluca commented:
> * Formatting / spacing on family modal seems incorrect compared to the other sections.

Will look into this 

> * Dragging the map around will place the marker, I don't believe it did this previously.

This has been corrected in the latest changes

> * Doesn't seem like there's any way to place the release marker on the map anymore? Maybe this still needs design work.

This has been updated in the new UI

> * Triggering error messages on the Add Telemetry modal will cause the dialog box to resize horizontally.

I'll look into this, will be part of the overall telemetry section changes coming

jeznorth committed:
Capture and Mortality UI Development

MacDeluca committed:
large refactor of telemetry related code

MacDeluca committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

al-rosenthal committed:
SIMSBIOHUB-354-355: Study area and Sampling sites on observations map (#1157)

SIMSBIOHUB-354-355: Added sampling site and study area features to the observation map on the survey details page
---------

Co-authored-by: Nick Phura <nickphura@gmail.com>

jeznorth committed:
UI WIP

MacDeluca committed:
continued telemetry refactor

MacDeluca committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

curtisupshall committed:
SIMSBIOHUB-348: Delete single or many observation records (#1154)

* Added a new ObservationTableContext
* Added an endpont and repository methods to delete one or more observation records
* Added loading button props to the YesNoDialog
* Refactored the observation context to support deleting one or more observation records

MacDeluca committed:
updated validation logic

jeznorth committed:
Small UI Updates / Fixes

jeznorth committed:
UI Updates

MacDeluca committed:
telemetry section stable

MacDeluca committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

MacDeluca committed:
format fix

MacDeluca committed:
added empty state for no animals

jeznorth committed:
Fixing deprecated component

GrahamS-Quartech committed:
SIMSBIOHUB-363: Import Survey Area Name and Description from Shapefile (#1161)

* Moved the shapefile name and description extraction to UI. Modified the sampling site POST api to accept custom names and desc from UI. Implemented the same ability in study area submission, now both parts of the form can reference the same logic for extracting this info.

GrahamS-Quartech commented:
Figured I would do another quick run through of this. 
- Seems like clicking on the map to place the pin and resizing the uncertainty does not work. Perhaps this is intentional?
- Clicking save on the animal subsection modals does not appear to fire a POST request right now.
- Mortality has a header for release location.
- Selecting Predation as reason allows you to select a taxon, then if you select a taxon and change the reason, it will grey out the taxon dropdown but not clear it. This behaviour might be fine though, I'm pretty sure the API will ignore the taxon in this case.
- May want to check in with Jeremy about integrating the 'View Family Structure' modal better.
- I think the lack of spacing between the critter ID and the general info card looks a bit odd.
- Telemetry validation seems a bit bugged but I probably just don't have latest fixes for that right now.

MacDeluca committed:
fixed main issue with nested formik forms / updated capture / mortality styling

MacDeluca committed:
Merge branch 'SIMSBIOHUB-188-Manage-Animals' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-188-Manage-Animals

MacDeluca committed:
commit before removing deprecated telemetry code

MacDeluca committed:
format fix

MacDeluca committed:
updated state for submission of dialogs

JeremyQuartech commented on app/src/constants/i18n.ts:
```diff
@@ -341,8 +341,10 @@ export const SurveyAnimalsI18N = {
     'The species or taxon of the animal. If the species is unknown, select the lowest-ranking known taxon, such as the genus or family.',
```
Typo: 'Different species may different units and unit names' should read 'Different species *may* have...'

MacDeluca replied:
fixed

MacDeluca committed:
removed old deprecated code for animal menu

JeremyQuartech reviewed (commented):
A couple things I noticed...

- The mortality workflow saving was unreliable. I tried to make a new mortality twice, and didn't get any confirmation from either. Had to refresh page and try again. The first one I added eventually showed up.
- Deleting a mortality event led to a runtime error, however, the deletion does succeed: `ERROR
Cannot read properties of undefined (reading 'projection_mode')`
- The capture workflow took a very long time to validate, but did succeed ultimately
- Deleting a capture event also is also causing the same runtime error as above
- Deleting a measurement causes the measurement dialog to briefly open and ten disappear

MacDeluca committed:
updated styling for capture form

GrahamS-Quartech commented:
- Editing a qualitative measurement and changing it to a different qualitative measurement type works correctly, but changing it to a quantitative measurement does not. I think this is because the resulting quantitative measurement object is placed in the update payload, but it probably needs to go in the create payload instead.
- I'm not sure if we should allow creation of multiple mortality events. This used to be disallowed when it was one big form.
- It seems like the validation for telemetry devices may still be firing unreliably. See screenshot, it's verifying against device id 1 even though I entered 456.
![image](https://github.com/bcgov/biohubbc/assets/112989452/925c3fe2-11ba-45f6-becf-e9e7810d1843)
- I was able to save and edit all other sections successfully.

jeznorth committed:
SIMSBIOHUB-334 UI Fixes (#1164)

* Minor UI polish and updating notification content.

JeremyQuartech committed:
SIMSBIOHUB-297: Keycloak Value Case Sensitivity (#1166)

* Removes toLowerCase casting on user_identifier fields from keycloak and stores raw string value in the database regardless of case
* Adds LOWER() sql statements on any relation to user_identifiers and user guids as to make them case-insensitive

MacDeluca committed:
created boilerplate for general/summary animal component

MacDeluca commented:
> A couple things I noticed...
> 
> * The mortality workflow saving was unreliable. I tried to make a new mortality twice, and didn't get any confirmation from either. Had to refresh page and try again. The first one I added eventually showed up.
> * Deleting a mortality event led to a runtime error, however, the deletion does succeed: `ERROR Cannot read properties of undefined (reading 'projection_mode')`
> * The capture workflow took a very long time to validate, but did succeed ultimately
> * Deleting a capture event also is also causing the same runtime error as above
> * Deleting a measurement causes the measurement dialog to briefly open and ten disappear

Found the issue for this, was a bad piece of state inside the dialog modal. It was keeping it open while the form was submitting, and subsequently while the form was deleting entries.

I think this fix solves all these issues.

MacDeluca committed:
updated state for dialog and mortality section

MacDeluca committed:
updated telemetry validation for device_id

MacDeluca committed:
removing file

MacDeluca commented:
> * Editing a qualitative measurement and changing it to a different qualitative measurement type works correctly, but changing it to a quantitative measurement does not. I think this is because the resulting quantitative measurement object is placed in the update payload, but it probably needs to go in the create payload instead.
> * I'm not sure if we should allow creation of multiple mortality events. This used to be disallowed when it was one big form.
> * It seems like the validation for telemetry devices may still be firing unreliably. See screenshot, it's verifying against device id 1 even though I entered 456.
>   ![image](https://user-images.githubusercontent.com/112989452/281194108-925c3fe2-11ba-45f6-becf-e9e7810d1843.png)
> * I was able to save and edit all other sections successfully.

Fixed :)

MacDeluca committed:
removed deprecated TelemetryDeviceForm

KjartanE committed:
SIMSBIOHUB-345: Display Time in Sampling Periods (#1167)

Added time to periods in sampling sites

---------

Co-authored-by: Alfred Rosenthal <alfred.k.rosenthal@gmail.com>

GrahamS-Quartech committed:
SIMSBIOHUB-371: Update Seed to Create Sampling Site and Observation Records (#1165)

* Seed script will now inject a survey sampling site with a corresponding method, period, and three observations.

MacDeluca committed:
updated general animal summary styling

MacDeluca committed:
updating telemetry section with date pickers

MacDeluca committed:
fixed date picker temp

curtisupshall committed:
SIMSBIOHUB-344: Tech debt: Observation Table Species display (#1160)

* Adds the TaxonomyContext for caching taxonomy names from IDs
* Updates the skeleton loader on the observations table
* Adds backend caching for taxonomy name lookup
* Updates service and repo methods for fetching observations supplementary data
* Fixes a bug with updating the observation count upon deletion
* Fixes a bug where deleted observation records would reappear when going back to the survey page after deletion
* Adds a missing snackbar shown upon deletion
* Adds the count of observtions being deleted in the confirm deletion dialog

al-rosenthal committed:
SIMSBIOHUB-361: Remove all study areas button (#1163)

Added a delete all button to study areas, updated map UI, added layer colors

---------

Co-authored-by: jeznorth <twenty15creative@gmail.com>

NickPhura committed:
ignore-skip

MacDeluca committed:
updated validation for device_make

MacDeluca committed:
final tweaks for telemetry section

curtisupshall reviewed (commented):
I added a **Do Not Merge** flag to this PR, since you need to change your base branch from `dev-staging` to `dev` (and resolve the subsequent merge conflicts).

MacDeluca committed:
fixed merge conflicts with dev-staging
